### PR TITLE
Phase3

### PR DESCRIPTION
**Pull request by jypark14**

### CODE REVIEW

File: src/components/counter.js [hunk 1]
```diff
@@ -40,21 +34,37 @@ export class Learning_Project extends Component {
...
-        </View>
-        <View style={styles.counter_row}>
+        <Button
+          large
+          title="Add counter"
+          color="blue"
+          onPress={this.props.actions.newCounter}
+        />
+        {this.props.counterList.map((counter, index) => {
+          return (
```
Comment (RubenSandwich) [inline src/components/counter.js]: This whole render in the map should be broken out into its own component as it is really separate.

File: src/components/counter.js [hunk 2]
```diff
@@ -40,21 +34,37 @@ export class Learning_Project extends Component {
...
-          />
-        </View>
-        <View style={styles.counter_row}>
+        <Button
+          large
+          title="Add counter"
+          color="blue"
+          onPress={this.props.actions.newCounter}
+        />
+        {this.props.counterList.map((counter, index) => {
```
Comment (RubenSandwich) [inline src/components/counter.js]: Using an array.map to render lists renders the whole list at once. For really large datasets this will break. It is better to use a component to lazily loads the data when needed so as not to create memory pressure. I suggest using FlatList: https://facebook.github.io/react-native/docs/flatlist.html.

File: src/reducers/counter.js [hunk 3]
```diff
@@ -3,29 +3,52 @@
 import { Actions } from "../actions/counter";
 
 const counter = {
```
Comment (RubenSandwich) [inline src/reducers/counter.js:5]: Is this state now unused? I couldn't find it used in the code, if it's unused then it should be removed.

File: src/reducers/counter.js [hunk 4]
```diff
@@ -3,29 +3,52 @@
...
+      counterList2[action.key].count = newcount;
+      // const counterList2 = state.counterList.map(
+      //   (counter, index) => counter + (index === action.key)
+      //   // index = {key}
+      // );
+
+      return Object.assign({ counterList: counterList2 }, state);
+    }
+
+    // const counterList2 = state.counterList.map(
```
Comment (RubenSandwich) [inline src/reducers/counter.js]: I'm guessing this is left in because you had trouble figuring out map? No worries let's talk about it today. Map, filter and reduce can take a while to get your head around.

File: src/components/counter.js [hunk 5]
```diff
@@ -40,21 +34,37 @@ export class Learning_Project extends Component {
...
+                  this.forceUpdate();
+                }}
+              />
+              <Text style={styles.text}>{counter.count}</Text>
+              <Button
+                large
+                title="Decrement -"
+                onPress={() => {
+                  this.props.actions.counterDecrement(index);
+                  this.forceUpdate();
```
Comment (RubenSandwich) [inline src/components/counter.js]: I'm happy you found a solution to the rerender problem, but this is more of a bandage than an actual solution. Because it depends on counterDecrement being a synchronous action, otherwise when forceUpdate fires the state will still be the same. While forceUpdate does exist it should be avoided unless absolutely needed.

**Review by RubenSandwich — CHANGES_REQUESTED** [review]
Good attempt. I know this struggling sucks, but I'm sure you feel like you are learning a ton about how Redux and React work and hitting some of its edge cases. It seems to me that your newness with map, filter and reduce are really getting you stuck here. Let's talk about this today.

Please fix the commented issues.

Also, you didn't update your tests. They are failing because you changed your reducer without updating your tests.

File: __tests__/counter.js [hunk 6]
```diff
@@ -2,58 +2,96 @@
...
+      counterDecrement(1)
+    );
+    expect(decreasedCounterCount.counterList).toEqual([
+      { count: 1, id: "abcd" },
+      { count: 2, id: "efgh" }
+    ]);
+  });
+});
+
+describe("counter close test", () => {
```
Comment (RubenSandwich) [inline __tests__/counter.js]: Awesome job writing more tests! 👍

File: package.json [hunk 7]
```diff
@@ -11,9 +11,11 @@
     "react-native": "0.55.4",
     "react-native-elements": "^0.19.1",
     "react-native-material-design": "^0.3.7",
+    "react-native-uuid": "^1.4.9",
```
Comment (RubenSandwich) [inline package.json]: This dependency seems to be unused and should be removed.

File: __tests__/counter.js [hunk 8]
```diff
@@ -2,58 +2,96 @@
...
+import { Actions, closeCounter } from "../src/actions/counter";
+import { counterCountReducer } from "../src/reducers/counter";
 import {
   counterIncrement,
   counterDecrement,
-  counterIncrement2,
-  counterDecrement2
+  newCounter
 } from "../src/actions/counter";
-describe("Stateless Components Render", () => {
```
Comment (RubenSandwich) [inline __tests__/counter.js]: How come you removed the component render test?

File: src/components/counter.js [hunk 9]
```diff
@@ -3,93 +3,90 @@
...
+          onPress={() => {
+            this.props.actions.closeCounter(this.props.index);
+          }}
+        />
       </View>
     );
   }
 }
 
 const mapStateToProps = state => ({
```
Comment (RubenSandwich) [inline src/components/counter.js]: Components shouldn't know about Redux. So they shouldn't have a connect function. Remember to just pass the required actions and data through props.

File: src/containers/counter.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  counterIncrement,
+  counterDecrement,
+  newCounter,
+  closeCounter
+} from "../actions/counter";
+import CounterRow from "../components/counter";
+import uuid from "uuid";
+
+const styles = StyleSheet.create({
+  counter_row: {
```
Comment (RubenSandwich) [inline src/containers/counter.js]: The counter_row style seems to be unused.

File: src/containers/counter.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    flexDirection: "column"
+  }
+});
+
+class CounterList extends Component {
+  renderItem({ data, index }) {
+    return (
+      <CounterRow
+        index={index}
+        data={this.props.counterList}
```
Comment (RubenSandwich) [inline src/containers/counter.js]: Why not just pass the cell data directly instead of the whole list? Looking at the CounterRow component and the actions it fires it only seems to need to know about the cell data itself. It should also clean up the CounterRow code by no longer needing lines like this to get the count: `this.props.data[this.props.index].count`.

File: src/components/counter.js [hunk 12]
```diff
@@ -3,93 +3,90 @@
...
 import { Button } from "react-native-elements";
 import {
   counterIncrement,
   counterDecrement,
-  counterIncrement2,
-  counterDecrement2
+  newCounter,
+  closeCounter
 } from "../actions/counter";
+import uuid from "uuid";
```
Comment (RubenSandwich) [inline src/components/counter.js]: uuid seems to be unused in this file.

File: src/reducers/counter.js [hunk 13]
```diff
@@ -1,31 +1,53 @@
 // @ts-check
 
 import { Actions } from "../actions/counter";
+import uuid from "uuid";
```
Comment (RubenSandwich) [inline src/reducers/counter.js]: uuid seems to be unused in this file.

File: src/reducers/counter.js [hunk 14]
```diff
@@ -1,31 +1,53 @@
...
-        count: state.count - 1
+    case Actions.newCounter: {
+      const counterNew = Object.assign({}, counter, { id: action.id });
+      const newCounterList = [...state.counterList, counterNew];
+      return Object.assign({}, state, { counterList: newCounterList });
+    }
+
+    case Actions.closeCounter: {
+      const deletedList = state.counterList.filter((elem, index) => {
+        if (index !== action.key) {
```
Comment (RubenSandwich) [inline src/reducers/counter.js]: Now that each counter in the counterList has a unique ID, you should use that to identify and update it in your Actions and Reducers. Indexes are very fragile and should only be used when nothing else is available. Using the key should also simplify your counter component code as well. (As the index prop will no longer be necessary.)

**Review by RubenSandwich — CHANGES_REQUESTED** [review]
Just a few places where you can simplify your code even more. Great job writing more tests!

**Review by RubenSandwich — APPROVED** [review]
🎉
